### DISCUSSION TITLE
Question about list[A | B | C]

### DISCUSSION THREAD

**Discussion by s-banach** [microsoft/pyright#5064] [General]
Good morning,
I know there have been issues about this before, I'd just like to clarify.

I might naively type a function like
```python
def f(x: list[A | B | C]):
    ...
```
with the intention that users of my package could pass in `list[A]` if they wanted.
But they'll get a type error if they do this.  
The justification for treating this as an error is that lists are mutable.  
If the type checker treated `list[A]` as an instance of `list[A|B|C]`,
then the function `f` might modify `x` by adding elements of type `B` or `C` to it.

(1) Is the only correct annotation for `x` to include all 7 combinations of `A,B,C`?

(2) Is there any way to annotate `x` as immutable, so that users can pass in `list[A]` and the type checker can understand that it's safe to treat `list[A]` as `list[A|B|C]`?
If not, have there been any python PEPs to add immutable lists?

Thanks for your time.

**Comment by erictraut** [comment]
Converting to discussion since this isn't a bug report.

**Reply by erictraut** [reply]
If your function is guaranteed not to mutate the list, you should annotate the input parameter as `Sequence[A | B | C]`. A `Sequence` is effectively an immutable list. Please refer to [this documentation](https://microsoft.github.io/pyright/#/type-concepts?id=generic-types) for additional details.

**Reply by s-banach** [reply]
Thanks very much for the help.
